INTMDB-655: PAK Resource Updates + Doc Cleanup + Deprecation Warnings

martinstibbe opened the pull request:
## Description

PAK Resource Updates + Doc Cleanup + Deprecation Warnings

Link to any related issue(s):

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [x] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments

martinstibbe committed:
Add project_assignment feature to permit assigning multiple projects to an API key

martinstibbe committed:
Add deprecation

martinstibbe committed:
lint

martinstibbe committed:
Add examples for project and org API key

martinstibbe committed:
Add additional multi project example

martinstibbe committed:
Update versions.tf

andreaangiolillo commented on mongodbatlas/resource_mongodbatlas_project.go:
```diff
@@ -78,9 +78,10 @@ func resourceMongoDBAtlasProject() *schema.Resource {
+				Type:       schema.TypeSet,
+				Optional:   true,
+				Computed:   true,
+				Deprecated: "This parameter is deprecated,please use project API Key resource instead",
```
Could you add this message to the `provider.go` (similar to  [provider.go#L42](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/b4045e1c5f98432a1127495e8ef238fd092ab513/mongodbatlas/provider.go#L42))?
It would be great to have only one deprecation message that we can use in all the resources/data sources. Thanks!

martinstibbe replied:
Added deprecation message from provider.go

andreaangiolillo commented on mongodbatlas/resource_mongodbatlas_project_api_key.go:
```diff
@@ -97,6 +157,51 @@ func resourceMongoDBAtlasProjectAPIKeyRead(ctx context.Context, d *schema.Resour
+			}
+			break
+		}
+	}
```
this piece of code seems to be duplicated in other parts of this file. Could you extract it in a function?

martinstibbe replied:
Added function to simplify

martinstibbe committed:
Add deprecation message create function to remove duplicated code

martinstibbe committed:
lint

andreaangiolillo commented on mongodbatlas/resource_mongodbatlas_project.go:
```diff
@@ -78,9 +79,10 @@ func resourceMongoDBAtlasProject() *schema.Resource {
+				Type:       schema.TypeSet,
+				Optional:   true,
+				Computed:   true,
+				Deprecated: fmt.Sprintf(DeprecationMessage, "v1.12.0", "mongodbatlas_project_api_key"),
```
this deprecation message does not seem correct. Here we are deprecating a param but the resource is not deprecated. I would expect to see another deprecation message define in provider.go for deprecating a parameter. Am I missing something here?

martinstibbe replied:
ok will make one specific to this unique case as it is not parameter to parameter change it is parameter to different resource

andreaangiolillo replied:
The main reason to have `"this parameter is deprecated, please use project API Key resource instead"` defined in `provider.go` as a constant is to use the same message once we have another parameter to deprecate in another resource. Feel free to ignore this comment if you think it is not necessary, we can always address this later

martinstibbe replied:
Added new message

martinstibbe committed:
Add new deprecation message

martinstibbe committed:
Merge branch 'release-staging-v.1.10.0' into INTMDB-655

andreaangiolillo reviewed (approved):
LGTM. Thanks for all the changes

themantissa commented on website/docs/r/access_list_api_key.html.markdown:
```diff
@@ -3,7 +3,7 @@ layout: "mongodbatlas"
 sidebar_current: "docs-mongodbatlas-resource-access_list-api-key"
 description: |-
-    Provides an Access List API Key resource.
+    Creates the access list entries for the specified Atlas Organization API Key. Atlas resources require that all API requests originate from IP addresses on the API access list.
```
"Atlas resources require that all API requests originate from IP addresses on the API access list." is not actually true as written.  Access Lists are super confusing with the Admin API.  Basically 1) some endpoints require an access list (like backup) while others do not (like clusters) by **default**.  2) Users can select to require an API key access list in the org settings in the UI with the "Require IP Access List for the Atlas Administration API" setting.  So we should make this confusing nuance clear (Last I heard IAM was going to clean this up so @Zuhairahmed we might want to check in with Fuat on that)

Zuhairahmed replied:
@themantissa good news, feedback from Fuat is that The restriction about #1 will be removed in a month or so, which means we'll have no IP access list requirement by default on any resource. Created [DOCSP-30474](https://jira.mongodb.org/browse/DOCSP-30474) to flow these new updates into all impacted resources and data sources in our Terraform Docs as well. Suggest for now we can commit your suggestion and then when Atlas Admin API has been updated we can then update all our Terraform docs for this fix at that time together.

themantissa replied:
```suggestion
    Creates the access list entries for the specified Atlas Organization API Key. 
```

themantissa replied:
@Zuhairahmed that's great news!  My suggestion then is just to remove this caveat for now since it's not technically correct.   Writing out what is correct is a bit of work so we might want to do that as a follow up.

themantissa commented on examples/atlas-api-key/create-and-assign-pak-seperately/README.md:
```diff
@@ -0,0 +1,12 @@
+# MongoDB Atlas Provider -- Create and Assign PAK sepeerately
```
All three examples are really good but this whole area can be hard to understand if the UI is your main PAK experience.  In the UI one has organization keys and project keys - if one looks closely you can tell  all are really  org keys and are simply assigned to projects but that's not as clear as it is in the Admin API.  I think an intro README at the atlas-api-key folder level explaining this and why we provided 3 examples would be very useful.   It's not a blocker to this PR but I think would be a good follow up, perhaps for the docs team?  cc @Zuhairahmed

Zuhairahmed replied:
Added this atlas-api-key README file directly, feel free to share any additional comments @zach-carr and @themantissa!

themantissa commented on examples/atlas-api-key/create-and-assign-pak-seperately/main.tf:
```diff
@@ -0,0 +1,32 @@
+  role_names  = ["ORG_OWNER"]
+}
+
+# Assign Newly Created Programmatic API Key to Project 
```
Should we provide a warning that this way is being deprecated in the near future?

Zuhairahmed replied:
+1 to deprecation comment, would also specify that they are going away in v1.12.0 release (2 versions out)

themantissa replied:
Not changed but not blocking - would be good to add.

Zuhairahmed replied:
added!

themantissa commented on mongodbatlas/resource_mongodbatlas_project_api_key.go:
```diff
@@ -143,13 +225,53 @@ func resourceMongoDBAtlasProjectAPIKeyUpdate(ctx context.Context, d *schema.Reso
 		}
 	}
 
+	if d.HasChange("project_assignment") {
```
Should we add a test to cover the changed functionality?

Zuhairahmed replied:
just confirming that @martinstibbe committed a new _test.go file which should address this comment

themantissa replied:
TY!

themantissa commented on website/docs/d/access_list_api_key.html.markdown:
```diff
@@ -3,7 +3,7 @@ layout: "mongodbatlas"
 sidebar_current: "docs-mongodbatlas-datasource-access-list-api-key"
 description: |-
-    Provides an Access List API Key resource.
+    Displays the access list entries for the specified Atlas Organization API Key. Atlas resources require that all API requests originate from IP addresses on the API access list.
```
See comment for resource about access lists and the Admin API.

themantissa replied:
```suggestion
    Displays the access list entries for the specified Atlas Organization API Key. 
```
Making same suggestion as above.  Not blocking but nice to correct.

themantissa commented on website/docs/r/project_api_key.html.markdown:
```diff
@@ -3,7 +3,7 @@ layout: "mongodbatlas"
+    Creates and assigns the specified Atlas Organization API Key to the specified Project. Users with the Project Owner role in the project associated with the API key can use the organization API key to access the resources.
 ---
 
 # Resource: mongodbatlas_project_api_key
```
Do we want to add an example for project_assignment?

martinstibbe replied:
Added create-api-key-assign-to-multiple-projects

themantissa replied:
Thank you!

themantissa commented on website/docs/r/project_api_key.html.markdown:
```diff
@@ -24,7 +24,7 @@ resource "mongodbatlas_project_api_key" "test" {
 ## Argument Reference
 
-* `project__id` - Unique identifier for the project whose API keys you want to retrieve. Use the /orgs endpoint to retrieve all organizations to which the authenticated user has access.
+* `project__id` -Unique 24-hexadecimal digit string that identifies your project.
```
Add project_assignment?

martinstibbe replied:
Added

themantissa replied:
TY!

themantissa commented on mongodbatlas/resource_mongodbatlas_project.go:
```diff
@@ -78,9 +79,10 @@ func resourceMongoDBAtlasProject() *schema.Resource {
+				Type:       schema.TypeSet,
+				Optional:   true,
+				Computed:   true,
+				Deprecated: fmt.Sprintf(DeprecationMessageParameterToResource, "v1.12.0", "mongodbatlas_project_api_key"),
```
The docs for project should also be updated with a deprecation message.

martinstibbe replied:
Added

themantissa replied:
Thank you!

themantissa reviewed (changes requested):
Comments inline, thank you!

themantissa commented:
@Zuhairahmed are we going to have our docs team work on this upgrade guide part of the scope?

martinstibbe committed:
Bump terraform version

zach-carr commented on examples/atlas-api-key/create-api-key-assign-to-multiple-projects/variables.tf:
```diff
@@ -0,0 +1,14 @@
+  description = "Private API key to authenticate to Atlas"
+}
+variable "project_id" {
+  description = "Atlas project name"
```
Looks like it should be "id"
```suggestion
  description = "Atlas project id"
```

zach-carr commented on examples/atlas-api-key/create-api-key-assign-to-multiple-projects/variables.tf:
```diff
@@ -0,0 +1,14 @@
+  default     = ""
+}
+variable "additional_project_id" {
+  description = "Atlas project name"
```
Description needs an update
```suggestion
variable "additional_project_id" {
  description = "Atlas project id"
```

zach-carr commented on mongodbatlas/resource_mongodbatlas_project_api_key.go:
```diff
@@ -143,13 +225,53 @@ func resourceMongoDBAtlasProjectAPIKeyUpdate(ctx context.Context, d *schema.Reso
 	}
 
+	if d.HasChange("project_assignment") {
+		// get the current api_keys and the new api_keys with changes
```
Super minor, but just for consistency in the comments, these could all start with capitalized verbs
```suggestion
		// Getting the current api_keys and the new api_keys with changes
```

zach-carr commented on mongodbatlas/resource_mongodbatlas_project_api_key.go:
```diff
@@ -143,13 +225,53 @@ func resourceMongoDBAtlasProjectAPIKeyUpdate(ctx context.Context, d *schema.Reso
+		// get the current api_keys and the new api_keys with changes
+		newAPIKeys, changedAPIKeys, removedAPIKeys := getStateProjectAssignmentAPIKeys(d)
+
+		// adding new api_keys into the project
```
```suggestion
		// Adding new api_keys into the project
```

zach-carr commented on website/docs/d/access_list_api_key.html.markdown:
```diff
@@ -52,12 +52,12 @@ data "mongodbatlas_access_list_api_key" "test" {
 * `api_key_id` - (Required) Unique identifier for the Organization API Key for which you want to retrieve an access list entry.
 * 
--> **NOTE:** One of the following attributes must set: `cidr_block`  or `ip_address`.
+-> **NOTE:** One of the following attributes must set: `cidr_block`  or `ip_address` not both.
```
I think adding a comma to "...cidr_block  or ip_address`,` not both" makes this a little clearer, but I suggest rephrasing to get rid of the colon:
```suggestion
-> **NOTE:** You must set either the `cidr_block` attribute or the `ip_address` attribute. Don't set both.
```

zach-carr commented on website/docs/d/access_list_api_keys.html.markdown:
```diff
@@ -59,12 +59,12 @@ data "mongodbatlas_access_list_api_key" "test" {
 * `api_key_id` - (Required) Unique identifier for the Organization API Key for which you want to retrieve an access list entry.
 * 
--> **NOTE:** One of the following attributes must set: `cidr_block`  or `ip_address`.
+-> **NOTE:** One of the following attributes must set: `cidr_block`  or `ip_address` not both.
```
The comment above applies here, too

zach-carr commented on website/docs/r/access_list_api_key.html.markdown:
```diff
@@ -40,12 +40,12 @@ resource "mongodbatlas_access_list_api_key" "test" {
 * `api_key_id` - Unique identifier for the Organization API Key for which you want to create a new access list entry.
 
--> **NOTE:** One of the following attributes must set: `cidr_block`  or `ip_address`.
+-> **NOTE:** One of the following attributes must set: `cidr_block`  or `ip_address` but not both.
```
The comment about this note applies here

zach-carr commented on examples/atlas-api-key/create-and-assign-pak-seperately/README.md:
```diff
@@ -0,0 +1,12 @@
+# MongoDB Atlas Provider -- Create and Assign PAK sepeerately
```
[typo]
```suggestion
# MongoDB Atlas Provider -- Create and Assign PAK Separately
```

zach-carr commented on examples/atlas-api-key/create-and-assign-pak-seperately/README.md:
```diff
@@ -0,0 +1,12 @@
+# MongoDB Atlas Provider -- Create and Assign PAK sepeerately
+This example creates an organization API Key and access list showing how to attach cidr_block.
```
Small suggestion for clarity
```suggestion
This example creates an organization API Key and access list, showing how to attach a CIDR block.
```

zach-carr commented on examples/atlas-api-key/create-and-assign-pak-seperately/README.md:
```diff
@@ -0,0 +1,12 @@
+- `public_key`: Atlas public key
+- `private_key`: Atlas  private key
+
+For this example, we will setup organization api key and attach access list to it.
```
Editing for clarity and consistency 
```suggestion
In this example, we will set up an organization API key and attach an access list to it.
```

zach-carr commented on examples/atlas-api-key/create-and-assign-pak-together/README.md:
```diff
@@ -0,0 +1,12 @@
+- `public_key`: Atlas public key
+- `private_key`: Atlas  private key
+
+For this example, we will setup project api key and attach access list to it.
```
Suggestion copied from above
```suggestion
This example creates a project API Key and access list, showing how to attach a CIDR block.

Variables Required to be set:
- `project_id`: ID of the Atlas project
- `org_id`: ID of Atlas organization
- `public_key`: Atlas public key
- `private_key`: Atlas  private key

In this example, we will set up a project API key and attach an access list to it.
```

zach-carr commented on examples/atlas-api-key/create-api-key-assign-to-multiple-projects/README.md:
```diff
@@ -0,0 +1,12 @@
+- `private_key`: Atlas  private key
+- `additional_project_id`: Additional project used to illustrate assign the API key to more than one project at a time
+
+For this example, we will setup project api key and attach it to multiple projects.
```
Similar suggestions
```suggestion
This example creates a project API key and assigns it to multiple projects.

Variables Required to be set:
- `project_id`: ID of the Atlas project that created the API key
- `public_key`: Atlas public key
- `private_key`: Atlas  private key
- `additional_project_id`: Additional project ID, used to assign the API key to more than one project at a time

In this example, we will set up a project API key and attach it to multiple projects.
```

zach-carr reviewed (changes requested):
Lots of very minor suggestions (sorry 😅), but I think all together they'll improve the flow of the page

martinstibbe committed:
Update project_api_key.html.markdown

martinstibbe committed:
Update project.html.markdown

martinstibbe committed:
Update docs

Zuhairahmed commented on examples/atlas-api-key/create-and-assign-pak-seperately/README.md:
```diff
@@ -0,0 +1,12 @@
+# MongoDB Atlas Provider -- Create and Assign PAK Separately
```
```suggestion
# MongoDB Atlas Provider -- Create and Assign PAK Separately (Deprecated) 

Note this functionality has been deprecated and will be removed in v1.12.0. Instead suggest see examples for Create and Assign PAK Together + Create and Assign PAK to Multiple Projects.
```

martinstibbe committed:
Add validation of project_assignment parameter

Zuhairahmed committed:
Create README.md

Zuhairahmed committed:
Update README.md

Zuhairahmed commented on examples/atlas-api-key/README.md:
```diff
@@ -0,0 +1,16 @@
+# MongoDB Atlas Provider -- MongoDB Atlas Programmatic API Key Examples 
```
Added this atlas-api-key README file directly (per below feedback), feel free to share any additional comments @zach-carr and @themantissa!

themantissa replied:
Not blocking but I think it would be helpful to explain the org/project API keys as I think that can still be confusing when using the API vs UI - i.e. All PAKs are organization API keys (org member role) but can be assigned to 1 or more projects in the org (such as for project owner role).

Zuhairahmed replied:
added!

Zuhairahmed committed:
Create PAK-upgrade-guide-1.10.0.html.markdown

Zuhairahmed committed:
Merge branch 'INTMDB-655' of https://github.com/mongodb/terraform-provider-mongodbatlas into INTMDB-655

Zuhairahmed committed:
Update PAK-upgrade-guide-1.10.0.html.markdown

Zuhairahmed commented:
> @Zuhairahmed are we going to have our docs team work on this upgrade guide part of the scope?

Added additional commit to this PR with PAK Upgrade Guide to address [above comment](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/1208#issuecomment-1579419346). Feel free to review/layer in any additional comments @themantissa @zach-carr. 

Also separately we will have our usual v1.10.0 upgrade guide as well (will start on this after pre-release is out) which will include all updates as part of v1.10.0 release and will link to this dedicated PAK Upgrade Guide one.

Zuhairahmed commented on website/docs/r/project.html.markdown:
```diff
@@ -79,6 +79,8 @@ api_keys allows one to assign an existing organization programmatic API key to a
 
 * `api_key_id` - (Required) The unique identifier of the Programmatic API key you want to associate with the Project.  The Programmatic API key and Project must share the same parent organization.  Note: this is not the `publicKey` of the Programmatic API key but the `id` of the key. See [Programmatic API Keys](https://docs.atlas.mongodb.com/reference/api/apiKeys/) for more.
 
+**WARNING:** This parameter is deprecated, use `mongodbatlas_project_api_key`  will be removed in v1.12.0 release
```
```suggestion
**WARNING:** The `api_keys` parameter is deprecated and will be removed in v1.12.0 release from codebase. Use `mongodbatlas_project_api_key`  resource instead. 
```

themantissa replied:
@Zuhairahmed this will need updating on the project data sources as well.   Maybe do as a follow up?   I'm going to approve this to unblock.

Zuhairahmed committed:
Update project_api_key.html.markdown

Zuhairahmed reviewed (approved):
LGTM with a few minor comments. also added a few commits directly as well

zach-carr commented on examples/atlas-api-key/README.md:
```diff
@@ -0,0 +1,16 @@
+
+In the Terraform MongoDB Atlas Provider v1.8.0 release, we introduced support for MongoDB Atlas Programmatic API Keys (PAK). While this was an exciting development, we quickly realized that some of our customers were facing challenges in optimally leveraging this resource. In response to this, we initiated several enhancements as part of the v1.10.0 release to refine the user experience with this ...
+
+The most notable improvements among these as part of v1.10.0 release, was the deprecation of the `api_keys` parameter from the `mongodbatlas_project` resource. We also extended the functionality of the `mongodbatlas_project_api_key` resource by incorporating the `project_assignment` parameter. This enhancement removes the necessity for users to create multiple `mongodbatlas_project` resource bloc ...
```
suggestion, feel free to take or leave - I think this first sentence could be simplified to something like:
"The most notable enhancement among these was the deprecation of the `api_keys` parameter from the ..."

otherwise, just a tweak to what's there:
```suggestion
The most notable improvement among these as part of v1.10.0 release was the deprecation of the `api_keys` parameter from the `mongodbatlas_project` resource. We also extended the functionality of the `mongodbatlas_project_api_key` resource by incorporating the `project_assignment` parameter. This enhancement removes the necessity for users to create multiple `mongodbatlas_project` resource blocks just to assign keys. This update streamlines the process of assigning an API Key to multiple projects, making it less cumbersome and more manageable.
```

zach-carr reviewed (approved):
LGTM

andreaangiolillo reviewed (approved):
LGTM

themantissa reviewed (approved):
Nothing blocking this but a few of docs improvements we can make now or do next round (cc @Zuhairahmed).  Thank you for all the changes!

Zuhairahmed committed:
Update website/docs/r/access_list_api_key.html.markdown

Co-authored-by: Melissa Plunkett <themantissa@users.noreply.github.com>

Zuhairahmed committed:
Update website/docs/d/access_list_api_key.html.markdown

Co-authored-by: Melissa Plunkett <themantissa@users.noreply.github.com>

Zuhairahmed reviewed (changes requested):
@martinstibbe can you flow through all above doc edits? it's mostly committing most of edits suggested by Zach, Melissa, and myself above and few new ones. after that I can do one last check and then merge. thanks!

martinstibbe committed:
Update examples/atlas-api-key/create-and-assign-pak-together/README.md

Co-authored-by: zach-carr <54542042+zach-carr@users.noreply.github.com>

martinstibbe committed:
Update examples/atlas-api-key/create-api-key-assign-to-multiple-projects/README.md

Co-authored-by: zach-carr <54542042+zach-carr@users.noreply.github.com>

martinstibbe committed:
Update examples/atlas-api-key/README.md

Co-authored-by: zach-carr <54542042+zach-carr@users.noreply.github.com>

martinstibbe committed:
Update website/docs/r/project.html.markdown

Co-authored-by: Zuhair Ahmed <zuhair.ahmed@mongodb.com>

martinstibbe committed:
Update examples/atlas-api-key/create-and-assign-pak-seperately/README.md

Co-authored-by: zach-carr <54542042+zach-carr@users.noreply.github.com>

martinstibbe committed:
Update examples/atlas-api-key/create-and-assign-pak-seperately/README.md

Co-authored-by: Zuhair Ahmed <zuhair.ahmed@mongodb.com>

martinstibbe committed:
Update examples/atlas-api-key/create-and-assign-pak-seperately/README.md

Co-authored-by: zach-carr <54542042+zach-carr@users.noreply.github.com>

martinstibbe commented:
> @martinstibbe can you flow through all above doc edits? it's mostly committing most of edits suggested by Zach, Melissa, and myself above and few new ones. after that I can do one last check and then merge. thanks!

completed suggested changes

martinstibbe committed:
Add additional deprecation note

martinstibbe committed:
Update README.md

martinstibbe committed:
Merge branch 'release-staging-v.1.10.0' into INTMDB-655

martinstibbe committed:
Bump up SDK to v0.29.0

Zuhairahmed committed:
Update README.md

More details to Readme explaining that all API keys are Organization Keys to help users

Zuhairahmed reviewed (approved):
LGTM and confirming @themantissa that believe we captured all requested doc improvements as part of this PR

themantissa reviewed (approved):
Thank you for the fixes!  LGTM!
